go/libraries/doltcore/merge/type_compatibility_test.go

```diff
@@ -128,69 +144,161 @@ func TestDoltIsTypeChangeCompatible(t *testing.T) {
-			to:       abcEnum,
-			expected: false,
-		}, {
-			name:     "enum collation changes are incompatible",
-			from:     abcEnum,
-			to:       abcEnumCi,
-			expected: false,
-		}, {
-			name:     "additive set changes are compatible",
-			from:     abcSet,
-			to:       abcdSet,
-			expected: true,
-		}, {
-			name:     "subtractive set changes are incompatible",
-			from:     abcdSet,
-			to:       abcSet,
-			expected: false,
-		}, {
-			name:     "set collation changes are incompatible",
-			from:     abcSet,
-			to:       abcSetCi,
-			expected: false,
-		}, {
-			name:     "geometry: identical types are compatible",
-			from:     geo,
-			to:       geo,
-			expected: true,
-		}, {
-			name:     "geometry: non-identical types are incompatible",
-			from:     geo,
-			to:       point,
-			expected: false,
+			name:       "equivalent types are compatible",
+			from:       typeinfo.Int64Type,
+			to:         typeinfo.Int64Type,
+			compatible: true,
+		}, {
+			name:       "int family: small to large type changes are incompatible",
+			from:       typeinfo.Int8Type,
+			to:         typeinfo.Int16Type,
+			compatible: false,
+		}, {
+			name:       "int family: large to small type changes are incompatible",
+			from:       typeinfo.Int64Type,
+			to:         typeinfo.Int16Type,
+			compatible: false,
+		}, {
+			name:       "additive enum changes are compatible",
+			from:       abcEnum,
+			to:         abcdEnum,
+			compatible: true,
+		}, {
+			name:       "subtractive enum changes are incompatible",
+			from:       abcdEnum,
+			to:         abcEnum,
+			compatible: false,
+		}, {
+			name:       "enum collation changes are incompatible",
+			from:       abcEnum,
+			to:         abcEnumCi,
+			compatible: false,
+		}, {
+			name:       "additive set changes are compatible",
+			from:       abcSet,
+			to:         abcdSet,
+			compatible: true,
+		}, {
+			name:       "subtractive set changes are incompatible",
+			from:       abcdSet,
+			to:         abcSet,
+			compatible: false,
+		}, {
+			name:       "set collation changes are incompatible",
+			from:       abcSet,
+			to:         abcSetCi,
+			compatible: false,
+		}, {
+			name:       "geometry: identical types are compatible",
+			from:       geo,
+			to:         geo,
+			compatible: true,
+		}, {
+			name:       "geometry: non-identical types are incompatible",
+			from:       geo,
+			to:         point,
+			compatible: false,
+		},
+
+		// Type width changes
+		{
+			name:       "type widening: varchar(10) to varchar(20)",
+			from:       varchar10,
+			to:         varchar20,
+			compatible: true,
+			rewrite:    false,
+		}, {
+			name:       "type narrowing: varchar(20) to varchar(10)",
+			from:       varchar20,
+			to:         varchar10,
+			compatible: false,
+		}, {
+			name:       "type widening: varchar to TEXT",
+			from:       varchar10,
+			to:         text,
+			compatible: true,
+			rewrite:    true,
+		}, {
+			name:       "type narrowing: TEXT to varchar",
```

Comment:
Seems wrong, need to rewrite the table in this case

end-to-end test would be nice too